pkg/eks/eks.go

```diff
@@ -228,18 +230,124 @@ func (c *ClusterProvider) WaitForControlPlane(id *api.ClusterMeta, clientSet *ku
 	}
 }
 
+type updateClusterConfigTask struct {
+	info string
+	spec *api.ClusterConfig
+	call func(*api.ClusterConfig) error
+}
+
+func (t *updateClusterConfigTask) Describe() string { return t.info }
+func (t *updateClusterConfigTask) Do(errs chan error) error {
+	err := t.call(t.spec)
+	close(errs)
+	return err
+}
+
+// GetCurrentClusterConfigForLogging fetches current cluster logging configuration as two sets - dsiable and enanaled types
+func (c *ClusterProvider) GetCurrentClusterConfigForLogging(cl *api.ClusterMeta) (sets.String, sets.String, error) {
+	enabled := sets.NewString()
+	disabled := sets.NewString()
+
+	cluster, err := c.DescribeControlPlaneMustBeActive(cl)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "unable to retrieve current cluster logging configuration")
+	}
+
+	for _, i := range cluster.Logging.ClusterLogging {
+		for _, t := range i.Types {
+			if t == nil {
+				return nil, nil, fmt.Errorf("unexpected response from EKS API - nil string")
+			}
+			if api.IsEnabled(i.Enabled) {
+				enabled.Insert(*t)
+			}
+			if api.IsDisabled(i.Enabled) {
+				disabled.Insert(*t)
+			}
+		}
+	}
+	return enabled, disabled, nil
+}
+
+// UpdateClusterConfigForLogging calls UpdateClusterConfig to enable logging
+func (c *ClusterProvider) UpdateClusterConfigForLogging(cfg *api.ClusterConfig) error {
+	enabled := sets.NewString()
+	disabled := sets.NewString()
+	if cfg.CloudWatch != nil && cfg.CloudWatch.ClusterLogging != nil {
+		enabled.Insert(cfg.CloudWatch.ClusterLogging.EnableTypes...)
+	}
+
+	for _, logType := range api.SupportedCloudWatchClusterLoggingTypes() {
+		if !enabled.Has(logType) {
+			disabled.Insert((logType))
+		}
+	}
+
+	input := &awseks.UpdateClusterConfigInput{
+		Name: &cfg.Metadata.Name,
+		Logging: &awseks.Logging{
+			ClusterLogging: []*awseks.LogSetup{
+				{
+					Enabled: api.Enabled(),
+					Types:   aws.StringSlice(enabled.List()),
+				},
+				{
+					Enabled: api.Disabled(),
+					Types:   aws.StringSlice(disabled.List()),
+				},
+			},
+		},
+	}
+
+	output, err := c.Provider.EKS().UpdateClusterConfig(input)
+	if err != nil {
+		return err
+	}
+	if err := c.waitForUpdateToSucceed(cfg.Metadata.Name, output.Update); err != nil {
```

Comment:
This is a bit odd: the updateClusterConfig operation is part of `c.Provider.EKS()` but the functionality that waits for it to complete is in the `c`. I think they should be both part of the same API (either EKS() or the `c`), otherwise it seems that there is not a clear responsibility boundary.

Reply:
`c.Provider.EKS()` is just a copy of the EKS client which `c` wraps. There are various places we need consume some of the clients (e.g. `./pkg/vpc`, `./pkg/iam`), that's why clients are exported and not completely hidden by the wrapper. If you have an idea of how we can clean this up, let's open another issue and discuss it there :)